forc-util/src/fs_locking.rs

```diff
@@ -0,0 +1,143 @@
+            return false;
+        }
+        System::new_all().process(Pid::from(pid)).is_some()
+    }
+
+    pub fn remove(&self) -> io::Result<()> {
+        if self.is_locked()? {
+            Err(io::Error::new(
+                std::io::ErrorKind::Other,
+                "Cannot remove a dirty lock file, it is locked by another process",
+            ))
+        } else {
+            self.remove_file()
+        }
+    }
+
+    fn remove_file(&self) -> io::Result<()> {
+        match remove_file(&self.0) {
+            Err(error) => {
+                if error.kind() == io::ErrorKind::NotFound {
+                    Ok(())
+                } else {
+                    Err(error)
+                }
+            }
+            _ => Ok(()),
+        }
+    }
+
+    pub fn is_locked(&self) -> io::Result<bool> {
+        let fs = File::open(&self.0);
+        println!("{:#?}", fs);
+        match fs {
+            Ok(mut file) => {
+                let mut pid = String::new();
+                file.read_to_string(&mut pid)?;
+                let is_locked = pid
+                    .trim()
+                    .parse::<usize>()
+                    .map(Self::is_pid_active)
+                    .unwrap_or_default();
+                drop(file);
+                if !is_locked {
+                    self.remove_file()?;
+                }
+                Ok(is_locked)
+            }
+            Err(err) => {
+                if err.kind() == io::ErrorKind::NotFound {
+                    Ok(false)
+                } else {
+                    Err(err)
+                }
+            }
+        }
+    }
+
+    pub fn lock(&self) -> io::Result<()> {
+        self.remove()?;
+        if let Some(dir) = self.0.parent() {
+            // Ensure the directory exists
+            create_dir_all(dir)?;
+        }
+
+        let mut fs = File::create(&self.0)?;
+        fs.write_all(std::process::id().to_string().as_bytes())?;
+        fs.sync_all()?;
+        fs.flush()?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::PidFileLocking;
+    use std::{
+        fs::{metadata, File},
+        io::{ErrorKind, Write},
+    };
+
+    #[test]
+    fn same_process() {
+        let x = PidFileLocking::lsp("test");
+        assert!(x.lock().is_ok());
+        // The current process is locking "test"
+        let x = PidFileLocking::lsp("test");
+        assert!(!x.is_locked().unwrap());
+    }
+
+    #[test]
+    fn stale() {
+        let x = PidFileLocking::lsp("stale");
+        assert!(x.lock().is_ok());
+
+        // lock file exists,
+        assert!(metadata(&x.0).is_ok());
+
+        // simulate a stale lock file
+        let mut x = File::create(&x.0).unwrap();
+        x.write_all(b"191919191919").unwrap();
+        x.flush().unwrap();
+        drop(x);
+
+        // PID=191919191919 does not exists, hopefully, and this should remove the lock file
+        let x = PidFileLocking::lsp("test");
```

Comment:
Shouldn't this use the same file name ("stale") to test that it's overwriting the stale file?